include/ruby/3/fl_type.h

```diff
@@ -0,0 +1,469 @@
+#define OBJ_FREEZE      RB_OBJ_FREEZE
+#define OBJ_FREEZE_RAW  RB_OBJ_FREEZE_RAW
+#define OBJ_FROZEN      RB_OBJ_FROZEN
+#define OBJ_FROZEN_RAW  RB_OBJ_FROZEN_RAW
+#define OBJ_INFECT      RB_OBJ_INFECT
+#define OBJ_INFECT_RAW  RB_OBJ_INFECT_RAW
+#define OBJ_TAINT       RB_OBJ_TAINT
+#define OBJ_TAINTABLE   RB_OBJ_TAINTABLE
+#define OBJ_TAINTED     RB_OBJ_TAINTED
+#define OBJ_TAINTED_RAW RB_OBJ_TAINTED_RAW
+#define OBJ_TAINT_RAW   RB_OBJ_TAINT_RAW
+#define OBJ_UNTRUST     RB_OBJ_UNTRUST
+#define OBJ_UNTRUSTED   RB_OBJ_UNTRUSTED
+/** @} */
+
+/* This is an enum because GDB wants it (rather than a macro) */
+enum { RUBY_FL_USHIFT = 12 };
+
+/* > The expression that defines the value  of an enumeration constant shall be
+ * > an integer constant expression that has a value representable as an `int`.
+ *
+ * -- ISO/IEC 9899:2018 section 6.7.2.2
+ *
+ * So ENUM_OVER_INT situation is an extension to the standard.  Note however
+ * that we do not support 16 bit `int` environment. */
+RB_GNUC_EXTENSION
+enum
+RUBY3_ATTR_FLAG_ENUM()
+ruby_fl_type {
+    RUBY_FL_WB_PROTECTED = (1<<5),
+    RUBY_FL_PROMOTED0    = (1<<5),
+    RUBY_FL_PROMOTED1    = (1<<6),
+    RUBY_FL_PROMOTED     = RUBY_FL_PROMOTED0 | RUBY_FL_PROMOTED1,
+    RUBY_FL_FINALIZE     = (1<<7),
+    RUBY_FL_TAINT        = (1<<8),
+    RUBY_FL_UNTRUSTED    = RUBY_FL_TAINT,
+    RUBY_FL_SEEN_OBJ_ID  = (1<<9),
+    RUBY_FL_EXIVAR       = (1<<10),
+    RUBY_FL_FREEZE       = (1<<11),
+
+#define RUBY3_FL_USER_N(n) RUBY_FL_USER##n = (1<<(RUBY_FL_USHIFT+n))
+    RUBY3_FL_USER_N(0),
+    RUBY3_FL_USER_N(1),
+    RUBY3_FL_USER_N(2),
+    RUBY3_FL_USER_N(3),
+    RUBY3_FL_USER_N(4),
+    RUBY3_FL_USER_N(5),
+    RUBY3_FL_USER_N(6),
+    RUBY3_FL_USER_N(7),
+    RUBY3_FL_USER_N(8),
+    RUBY3_FL_USER_N(9),
+    RUBY3_FL_USER_N(10),
+    RUBY3_FL_USER_N(11),
+    RUBY3_FL_USER_N(12),
+    RUBY3_FL_USER_N(13),
+    RUBY3_FL_USER_N(14),
+    RUBY3_FL_USER_N(15),
+    RUBY3_FL_USER_N(16),
+    RUBY3_FL_USER_N(17),
+    RUBY3_FL_USER_N(18),
+#if ENUM_OVER_INT
+    RUBY3_FL_USER_N(19),
+#else
+# define RUBY_FL_USER19 (RUBY3_VALUE_ONE<<(RUBY_FL_USHIFT+19))
+#endif
+#undef RUBY3_FL_USER_N
+#undef RUBY3_WIDER_ENUM
+
+    RUBY_ELTS_SHARED  = RUBY_FL_USER2,
+    RUBY_FL_SINGLETON = RUBY_FL_USER0,
+};
+
+enum { RUBY_FL_DUPPED = RUBY_T_MASK | RUBY_FL_EXIVAR | RUBY_FL_TAINT };
+
+RUBY3_SYMBOL_EXPORT_BEGIN()
+void rb_obj_infect(VALUE victim, VALUE carrier);
+void rb_freeze_singleton_class(VALUE klass);
+RUBY3_SYMBOL_EXPORT_END()
+
+RUBY3_ATTR_PURE_ON_NDEBUG()
+RUBY3_ATTR_ARTIFICIAL()
+RUBY3_ATTR_FORCEINLINE()
+static bool
+RB_FL_ABLE(VALUE obj)
+{
+    if (RB_SPECIAL_CONST_P(obj)) {
+        return false;
+    }
+    else if (RB_TYPE_P(obj, RUBY_T_NODE)) {
+        return false;
+    }
+    else {
+        return true;
+    }
+}
+
+RUBY3_ATTR_PURE_ON_NDEBUG()
+RUBY3_ATTR_ARTIFICIAL()
+static inline VALUE
+RB_FL_TEST_RAW(VALUE obj, VALUE flags)
+{
+    RUBY3_ASSERT_OR_ASSUME(RB_FL_ABLE(obj));
```

Comment:
I guess this is a new assertion introduced in this PR, and I'd like to share this helped to find https://github.com/ruby/ruby/commit/04e56958e6a3c3278d0d3a087954ee8aca2fbd10. Thank you!